grpc/runtime/src/main/scala/io/buoyant/grpc/runtime/EventStream.scala

```diff
@@ -0,0 +1,133 @@
+package io.buoyant.grpc.runtime
+
+import com.twitter.util.{Activity, Event, Future, Promise, Return, Throw, Try, Var}
+import java.util.concurrent.atomic.AtomicReference
+import scala.annotation.tailrec
+
+object EventStream {
+
+  def apply[T](event: Event[Ev[T]]): Stream[T] =
+    new EventStream(event)
+
+  def apply[T](var0: Var[Ev[T]]): Stream[T] =
+    apply(var0.changes)
+
+  def apply[T](act: Activity[T]): Stream[T] =
+    apply(act.values.map {
+      case Return(t) => Val(t)
+      case Throw(e) => End(Throw(e))
+    })
+
+  /** An T-typed event value */
+  sealed trait Ev[+T]
+
+  /** A non-terminal event, containing a value */
+  case class Val[T](value: T) extends Ev[T]
+
+  /** A terminal event, containing either a value or an error */
+  case class End[T](value: Try[T]) extends Ev[T]
+
+  private sealed trait State[+T]
+  private object State {
+    /** There are no unconsumed values and no waiting receivers. */
+    object Empty extends State[Nothing]
+
+    /** There are no unconsumed values and there is a waiting receiver. */
+    case class Recving[T](onUpdate: Promise[Unit]) extends State[T]
+
+    /** An unconsumed value is ready. */
+    case class Updated[T](value: Ev[T]) extends State[T]
+
+    /** The event is no longer being observed and no further values will be produced. */
+    object Closed extends State[Nothing]
+  }
+}
+
+/**
+ * A gRPC stream for com.twitter.util.Event types.
+ *
+ * Because this is designed to be used in the context of an
+ * Activity/Var, the stream may NOT reflect all event states.
+ * Intermediate state transitions may be dropped in favor of buffering
+ * only the last-observed state.
+ *
+ * Event values are wrapped by EventStream.Ev to indicate whether a
+ * value is terminal.
+ *
+ * TODO streams are not properly failed on event failure.
+ *
+ * TODO when a stream fails (from the remote), we should eagerly
+ * stop observing the event.
+ */
+class EventStream[+T](events: Event[EventStream.Ev[T]]) extends Stream[T] {
```

Comment:
Just wondering aloud if we should name this in a way to indicate it does not reflect all event states.  LossyEventStream?  EventuallyConsistentEventStream?